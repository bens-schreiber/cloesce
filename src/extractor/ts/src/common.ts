export type Either<L, R> = { ok: false; value: L } | { ok: true; value: R };
export function left<L>(value: L): Either<L, never> {
  return { ok: false, value };
}
export function right<R>(value: R): Either<never, R> {
  return { ok: true, value };
}

export enum ExtractorErrorCode {
  UnknownType,
  MultipleGenericType,
  InvalidIncludeTree,
  UnknownNavigationPropertyReference,
  InvalidNavigationPropertyReference,
  MissingNavigationPropertyReference,
  MissingManyToManyUniqueId,
  MissingPrimaryKey,
  MissingWranglerEnv,
  TooManyWranglerEnvs,
  TooManyMiddlewares,
  MissingMiddlewareMethod,
}

const errorInfoMap: Record<
  ExtractorErrorCode,
  { description: string; suggestion: string }
> = {
  [ExtractorErrorCode.UnknownType]: {
    description: "Encountered an unknown or unsupported type",
    suggestion: "Refer to the documentation on valid Cloesce TS types",
  },
  [ExtractorErrorCode.MultipleGenericType]: {
    description: "Cloesce does not yet support types with multiple generics",
    suggestion:
      "Simplify your type to use only a single generic parameter, ie Foo<T>",
  },
  [ExtractorErrorCode.InvalidIncludeTree]: {
    description: "Invalid Include Tree",
    suggestion:
      "Include trees must only contain references to a model's navigation properties.",
  },
  [ExtractorErrorCode.UnknownNavigationPropertyReference]: {
    description: "Unknown Navigation Property Reference",
    suggestion:
      "Verify that the navigation property reference model exists, or create a model.",
  },
  [ExtractorErrorCode.InvalidNavigationPropertyReference]: {
    description: "Invalid Navigation Property Reference",
    suggestion: "Ensure the navigation property points to a valid model field",
  },
  [ExtractorErrorCode.MissingNavigationPropertyReference]: {
    description: "Missing Navigation Property Reference",
    suggestion:
      "Navigation properties require a foreign key model attribute reference",
  },
  [ExtractorErrorCode.MissingManyToManyUniqueId]: {
    description: "Missing unique id on Many to Many navigation property",
    suggestion:
      "Define a unique identifier field for the Many-to-Many relationship",
  },
  [ExtractorErrorCode.MissingPrimaryKey]: {
    description: "Missing primary key on a model",
    suggestion: "Add a primary key field to your model (e.g., `id: number`)",
  },
  [ExtractorErrorCode.MissingWranglerEnv]: {
    description: "Missing a wrangler environment definition in the project",
    suggestion: "Add a @WranglerEnv class in your project.",
  },
  [ExtractorErrorCode.TooManyWranglerEnvs]: {
    description: "Too many wrangler environments defined in the project",
    suggestion: "Consolidate or remove unused @WranglerEnv's",
  },
  [ExtractorErrorCode.TooManyMiddlewares]: {
    description: "Too many middleware classes defined in the project",
    suggestion: "Only one @Middleware class is allowed per project",
  },
  [ExtractorErrorCode.MissingMiddlewareMethod]: {
    description: "Middleware class must have a 'handle' method",
    suggestion: "Add a 'handle' method to your @Middleware class",
  },
};

export function getErrorInfo(code: ExtractorErrorCode) {
  return errorInfoMap[code];
}

export class ExtractorError {
  context?: string;
  snippet?: string;

  constructor(public code: ExtractorErrorCode) {}

  addContext(fn: (val: string | undefined) => string | undefined) {
    this.context = fn(this.context ?? "");
  }
}

export type HttpResult<T = unknown> = {
  ok: boolean;
  status: number;
  data?: T;
  message?: string;
};

export type CidlType =
  | "Void"
  | "Integer"
  | "Real"
  | "Text"
  | "Blob"
  | { Inject: string }
  | { Object: string }
  | { Nullable: CidlType }
  | { Array: CidlType }
  | { HttpResult: CidlType };

export function isNullableType(ty: CidlType): boolean {
  return typeof ty === "object" && ty !== null && "Nullable" in ty;
}

export enum HttpVerb {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  PATCH = "PATCH",
  DELETE = "DELETE",
}

export interface NamedTypedValue {
  name: string;
  cidl_type: CidlType;
}

export interface ModelAttribute {
  value: NamedTypedValue;
  foreign_key_reference: string | null;
}

export interface ModelMethod {
  name: string;
  is_static: boolean;
  http_verb: HttpVerb;
  return_type: CidlType | null;
  parameters: NamedTypedValue[];
}

export type NavigationPropertyKind =
  | { OneToOne: { reference: string } }
  | { OneToMany: { reference: string } }
  | { ManyToMany: { unique_id: string } };

export interface NavigationProperty {
  var_name: string;
  model_name: string;
  kind: NavigationPropertyKind;
}

export function getNavigationPropertyCidlType(
  nav: NavigationProperty
): CidlType {
  return "OneToOne" in nav.kind
    ? { Object: nav.model_name }
    : { Array: { Object: nav.model_name } };
}

export interface Model {
  name: string;
  primary_key: NamedTypedValue;
  attributes: ModelAttribute[];
  navigation_properties: NavigationProperty[];
  methods: Record<string, ModelMethod>;
  data_sources: Record<string, DataSource>;
  source_path: string;
}

export interface PlainOldObject {
  name: string;
  attributes: NamedTypedValue[];
  source_path: string;
}

export interface CidlIncludeTree {
  [key: string]: CidlIncludeTree;
}

export interface DataSource {
  name: string;
  tree: CidlIncludeTree;
}

export interface WranglerEnv {
  name: string;
  source_path: string;
}

export interface MiddlewareMethod {
  name: string;
  parameters: NamedTypedValue[];
}

export interface Middleware {
  class_name: string;
  method: MiddlewareMethod;
  source_path: string;
}

export interface CloesceAst {
  version: string;
  project_name: string;
  language: "TypeScript";
  wrangler_env: WranglerEnv;
  models: Record<string, Model>;
  poos: Record<string, PlainOldObject>;
  middleware: Middleware | null;
}
