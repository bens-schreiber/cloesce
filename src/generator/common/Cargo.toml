[package]
name = "common"
version = "0.0.1"
edition = "2024"

[dependencies]
anyhow = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
serde_with = "3.14.1"
toml = { workspace = true }
