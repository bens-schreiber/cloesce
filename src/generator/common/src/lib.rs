pub mod builder;
pub mod err;

use std::collections::BTreeMap;
use std::path::PathBuf;

use err::GeneratorErrorKind;
use err::Result;
use serde::Deserialize;
use serde::Serialize;
use serde_with::MapPreventDuplicates;
use serde_with::serde_as;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, Hash)]
pub enum CidlType {
    /// No type
    Void,

    /// SQLite integer
    Integer,

    /// SQLite floating point number
    Real,

    /// SQLite string
    Text,

    /// SQLite large structured data
    Blob,

    /// A dependency injected instance, containing a type name.
    Inject(String),

    /// A model, or plain old object, containing the name of the class.
    Object(String),

    /// A part of a model or plain object, containing the name of the class.
    ///
    /// Only valid as a method argument.
    Partial(String),

    /// An array of any type
    Array(Box<CidlType>),

    /// A REST API response, which can contain any type or nothing.
    HttpResult(Box<CidlType>),

    /// A wrapper denoting the type within can be null.
    /// If the inner value is void, represents just null.
    Nullable(Box<CidlType>),
}

impl CidlType {
    /// Returns the inner part of an array if the type is an array
    pub fn unwrap_array(&self) -> Option<&CidlType> {
        match self {
            CidlType::Array(inner) => Some(inner),
            _ => None,
        }
    }

    /// Returns the root most CidlType, being any non Model/Array/Result.
    pub fn root_type(&self) -> &CidlType {
        match self {
            CidlType::Array(inner) => inner.root_type(),
            CidlType::HttpResult(inner) => inner.root_type(),
            CidlType::Nullable(inner) => inner.root_type(),
            t => t,
        }
    }

    pub fn is_nullable(&self) -> bool {
        matches!(self, CidlType::Nullable(_))
    }

    pub fn array(cidl_type: CidlType) -> CidlType {
        CidlType::Array(Box::new(cidl_type))
    }

    pub fn nullable(cidl_type: CidlType) -> CidlType {
        CidlType::Nullable(Box::new(cidl_type))
    }

    pub fn null() -> CidlType {
        CidlType::Nullable(Box::new(CidlType::Void))
    }
}

#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub enum HttpVerb {
    GET,
    POST,
    PUT,
    PATCH,
    DELETE,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Hash)]
pub struct NamedTypedValue {
    pub name: String,
    pub cidl_type: CidlType,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ModelAttribute {
    pub value: NamedTypedValue,
    pub foreign_key_reference: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ModelMethod {
    pub name: String,
    pub is_static: bool,
    pub http_verb: HttpVerb,
    pub return_type: CidlType,
    pub parameters: Vec<NamedTypedValue>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct IncludeTree(pub BTreeMap<String, IncludeTree>);

#[derive(Serialize, Deserialize, Debug)]
pub struct DataSource {
    pub name: String,
    pub tree: IncludeTree,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum NavigationPropertyKind {
    OneToOne { reference: String },
    OneToMany { reference: String },
    ManyToMany { unique_id: String },
}

#[derive(Serialize, Deserialize, Debug)]
pub struct NavigationProperty {
    pub var_name: String,
    pub model_name: String,
    pub kind: NavigationPropertyKind,
}

#[serde_as]
#[derive(Serialize, Deserialize, Debug)]
pub struct Model {
    pub name: String,
    pub primary_key: NamedTypedValue,
    pub attributes: Vec<ModelAttribute>,
    pub navigation_properties: Vec<NavigationProperty>,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub methods: BTreeMap<String, ModelMethod>,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub data_sources: BTreeMap<String, DataSource>,
    pub source_path: PathBuf,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PlainOldObject {
    pub name: String,
    pub attributes: Vec<NamedTypedValue>,
    pub source_path: PathBuf,
}

#[derive(Serialize, Deserialize)]
pub enum InputLanguage {
    TypeScript,
}

#[derive(Serialize, Deserialize)]
pub struct WranglerEnv {
    pub name: String,
    pub source_path: PathBuf,
}

#[serde_as]
#[derive(Serialize, Deserialize)]
pub struct CloesceAst {
    pub version: String,
    pub project_name: String,
    pub language: InputLanguage,
    pub wrangler_env: WranglerEnv,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub models: BTreeMap<String, Model>,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub poos: BTreeMap<String, PlainOldObject>,
}

impl CloesceAst {
    pub fn from_json(path: &std::path::Path) -> Result<CloesceAst> {
        let cidl_contents = std::fs::read_to_string(path).map_err(|e| {
            GeneratorErrorKind::InvalidInputFile
                .to_error()
                .with_context(e.to_string())
        })?;
        serde_json::from_str::<CloesceAst>(&cidl_contents).map_err(|e| {
            GeneratorErrorKind::InvalidInputFile
                .to_error()
                .with_context(e.to_string())
        })
    }

    fn is_valid_object_ref(&self, o: &str) -> bool {
        self.models.contains_key(o) || self.poos.contains_key(o)
    }

    /// Ensures all `CidlTypes` are logically correct for the area, essentially doing
    /// the first level of semantic analysis for the generator.
    ///
    /// Returns error on
    /// - Model attributes with invalid SQL types
    /// - Primary keys with invalid SQL types
    /// - Invalid Model or Method map K/V
    /// - Unknown navigation property model references
    /// - Unknown model references in method parameters
    /// - Invalid method parameter types
    pub fn validate_types(&self) -> Result<()> {
        let ensure_valid_sql_type = |model: &Model, value: &NamedTypedValue| {
            let inner = match &value.cidl_type {
                CidlType::Nullable(inner) if matches!(inner.as_ref(), CidlType::Void) => {
                    fail!(GeneratorErrorKind::NullSqlType)
                }
                CidlType::Nullable(inner) => inner.as_ref(),
                other => other,
            };

            ensure!(
                matches!(
                    inner,
                    CidlType::Integer | CidlType::Real | CidlType::Text | CidlType::Blob
                ),
                GeneratorErrorKind::InvalidSqlType,
                "{}.{}",
                model.name,
                value.name
            );

            Ok(())
        };

        for (model_name, model) in &self.models {
            // Validate record
            ensure!(
                *model_name == model.name,
                GeneratorErrorKind::InvalidMapping,
                "Model record key did not match it's model name? {} : {}",
                model_name,
                model.name
            );

            // Validate PK
            ensure_valid_sql_type(model, &model.primary_key)?;

            // Validate attributes
            for a in &model.attributes {
                ensure_valid_sql_type(model, &a.value)?;

                if let Some(fk_model) = &a.foreign_key_reference {
                    // Validate the fk's model exists
                    ensure!(
                        self.models.contains_key(fk_model.as_str()),
                        GeneratorErrorKind::UnknownObject,
                        "{}.{} => {}?",
                        model.name,
                        a.value.name,
                        fk_model
                    );
                }
            }

            // Validate navigation props
            for nav in &model.navigation_properties {
                ensure!(
                    self.models.contains_key(nav.model_name.as_str()),
                    GeneratorErrorKind::UnknownObject,
                    "{} => {}?",
                    model.name,
                    nav.model_name
                );
            }

            // Validate methods
            for (method_name, method) in &model.methods {
                // Validate record
                ensure!(
                    *method_name == method.name,
                    GeneratorErrorKind::InvalidMapping,
                    "Method record key did not match it's method name? {}: {}",
                    method_name,
                    method.name
                );

                // Validate return type
                match &method.return_type {
                    CidlType::Object(o) => {
                        ensure!(
                            self.is_valid_object_ref(o),
                            GeneratorErrorKind::UnknownObject,
                            "{}.{}",
                            model.name,
                            method.name
                        );
                    }
                    CidlType::Partial(_) => {
                        fail!(
                            GeneratorErrorKind::UnexpectedPartialReturn,
                            "{}.{}",
                            model.name,
                            method.name,
                        )
                    }
                    _ => {}
                }

                // Validate method params
                for param in &method.parameters {
                    let root_type = param.cidl_type.root_type();

                    match root_type {
                        CidlType::Void => {
                            fail!(
                                GeneratorErrorKind::UnexpectedVoid,
                                "{}.{}.{}",
                                model.name,
                                method.name,
                                param.name
                            )
                        }
                        CidlType::Object(o) | CidlType::Partial(o) => {
                            ensure!(
                                self.is_valid_object_ref(o),
                                GeneratorErrorKind::UnknownObject,
                                "{}.{}.{}",
                                model.name,
                                method.name,
                                param.name
                            );

                            // TODO: remove this
                            if method.http_verb == HttpVerb::GET {
                                fail!(
                                    GeneratorErrorKind::NotYetSupported,
                                    "GET Requests currently do not support model parameters {}.{}.{}",
                                    model.name,
                                    method.name,
                                    param.name
                                )
                            }
                        }
                        _ => {
                            // Ignore
                        }
                    }
                }
            }
        }

        Ok(())
    }
}
