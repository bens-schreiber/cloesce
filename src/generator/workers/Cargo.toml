[package]
name = "workers"
version = "0.0.1"
edition = "2024"

[dependencies]
anyhow = { workspace = true }
common = { path = "../common" }
client = { path = "../client" }
wrangler = { path = "../wrangler" }
pathdiff = { version = "0.2.3" }

[dev-dependencies]
serde_json = { workspace = true }
