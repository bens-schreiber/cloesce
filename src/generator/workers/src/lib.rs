mod typescript;
use common::{CidlSpec, HttpVerb, InputLanguage, Method, Model, TypedValue};
use typescript::TypescriptWorkersGenerator;

trait LanguageWorkerGenerator {
    fn imports(&self, models: &[Model]) -> String;
    fn preamble(&self) -> String;
    fn validators(&self, models: &[Model]) -> String;
    fn main(&self) -> String;
    fn router(&self, model: String) -> String;
    fn router_model(&self, model_name: &str, method: String) -> String;
    fn router_method(&self, method: &Method, proto: String) -> String;
    fn proto(&self, method: &Method, body: String) -> String;
    fn validate_http(&self, verb: &HttpVerb) -> String;
    fn validate_req_body(&self, params: &[TypedValue]) -> String;
    fn hydrate_model(&self, model_name: &Model) -> String;
    fn dispatch_method(&self, model_name: &str, method: &Method) -> String;
}

pub struct WorkersFactory {
    domain: Option<String>,
}

impl WorkersFactory {
    pub fn new(domain: String) -> Self {
        Self { domain: Some(domain) }
    }

    // If you want builder-style, you can still add a setter:
    pub fn with_domain(mut self, domain: String) -> Self {
        self.domain = Some(domain);
        self
    }
   
    fn model(model: &Model, lang: &dyn LanguageWorkerGenerator) -> String {
        let mut static_methods = vec![];
        let mut instance_methods = vec![];
       
        for method in &model.methods {
            let validate_http = lang.validate_http(&method.http_verb);
            let validate_params = lang.validate_req_body(&method.parameters);
            let hydration = if method.is_static {
                ""
            } else {
                &lang.hydrate_model(model)
            };
            let dispatch = lang.dispatch_method(&model.name, method);
            let method_body = format!(
                r#"
                {validate_http}
                {validate_params}
                {hydration}
                {dispatch}
            "#
            );
            let proto = lang.proto(method, method_body);
            let router_method = lang.router_method(method, proto);
           
            if method.is_static {
                static_methods.push(router_method);
            } else {
                instance_methods.push(router_method);
            }
        }
       
        // Combine static methods and instance methods under <id> if any exist
        let mut all_methods = static_methods;
        if !instance_methods.is_empty() {
            let instance_routes = instance_methods.join(",\n");
            all_methods.push(format!(r#""<id>": {{ {} }}"#, instance_routes));
        }
       
        lang.router_model(&model.name, all_methods.join(",\n"))
    }

    pub fn create(self, spec: CidlSpec) -> String {
        let generator: Box<dyn LanguageWorkerGenerator> = match spec.language {
            InputLanguage::TypeScript => {
                Box::new(TypescriptWorkersGenerator::new(self.domain))
            }
        };

        let imports = generator.imports(&spec.models);
        let preamble = generator.preamble();
        let validators = generator.validators(&spec.models);
        let router = {
            let router_body = spec
                .models
                .iter()
                .map(|m| Self::model(m, generator.as_ref()))
                .collect::<Vec<_>>()
                .join("\n");
            generator.router(router_body)
        };
        let main = generator.main();
       
        format!(
            r#"
{imports}
{preamble}
{validators}
{router}
{main}
        "#
        )
    }
}