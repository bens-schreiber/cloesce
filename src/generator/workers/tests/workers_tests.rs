use anyhow::Result;
use common::CidlSpec;
use insta::assert_snapshot;
use workers::WorkersFactory;
use std::path::PathBuf;

fn load_cidl() -> Result<CidlSpec> {
    let cidl_path = PathBuf::from("../fixtures/cidl.json");
    let cidl_contents = std::fs::read_to_string(cidl_path)?;
    Ok(serde_json::from_str::<CidlSpec>(&cidl_contents)?)
}

#[test]
fn test_generate_client_snapshot() -> Result<()> {
    // Arrange
    let cidl = load_cidl()?;
   
    // Use the default domain here
    let workers = WorkersFactory::new("localhost".to_string()).create(cidl);
   
    // Assert
    assert_snapshot!("generated_workers", workers);
    Ok(())
}

#[test]
fn test_generate_client_with_custom_domain() -> Result<()> {
    // Arrange
    let cidl = load_cidl()?;
   
    // Use a custom domain with path to test root extraction
    let workers = WorkersFactory::new("example.com/foo/bar/baz".to_string())
        .create(cidl);
   
    // Assert: router should use "baz" as root
    assert!(workers.contains("const router = { baz:"));
   
    Ok(())
}

#[test]
fn test_domain_normalization() -> Result<()> {
    // Test various domain formats
    let test_cases = vec![
        ("localhost:5000/myapi", "myapi"),
        ("example.com/v1/api", "api"),
        ("api.example.com/services", "services"),
        ("example.com", "api"), // no path, defaults to "api"
    ];
   
    for (domain, expected_root) in test_cases {
        // Arrange
        let cidl = load_cidl()?;
       
        // Act
        let workers = WorkersFactory::new(domain.to_string())
            .create(cidl);
       
        // Assert
        let expected = format!("const router = {{ {}:", expected_root);
        assert!(
            workers.contains(&expected),
            "Domain '{}' should produce root '{}', but got:\n{}",
            domain,
            expected_root,
            workers.lines()
                .find(|l| l.contains("const router"))
                .unwrap_or("router not found")
        );
    }
   
    Ok(())
}