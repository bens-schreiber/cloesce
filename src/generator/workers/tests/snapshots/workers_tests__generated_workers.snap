---
source: workers/tests/workers_tests.rs
expression: workers
---
import { D1Database } from "@cloudflare/workers-types"

import { ExampleModel } from '../example_model'; 

// @ts-nocheck
export interface Env {
  DB: D1Database;
}
type SQLRow = Record<string, any>;
/**
 * TODO: This could be WASM
 * TODO: This is all GPT slop
 *
 * @returns JSON of type T
 */
export function mapSql<T>(rows: SQLRow[]): T[] {
  const result: any[] = [];
  const entityMaps: Record<string, Map<string, any>> = {};
  function getEntityKey(entityObj: any): string {
    // Use JSON string of all values as a "unique key" for deduplication
    return JSON.stringify(entityObj);
  }
  for (const row of rows) {
    const rowEntities: Record<string, any> = {};
    // Step 1: split row into entities by prefix
    for (const col in row) {
      const parts = col.split("_");
      if (parts.length < 2) continue;
      const entity = parts[0];
      const field = parts.slice(1).join("_");
      if (!rowEntities[entity]) rowEntities[entity] = {};
      rowEntities[entity][field] = row[col];
    }
    // Step 2: merge entities into result
    let topObj: any = null;
    for (const entity in rowEntities) {
      const entityObj = rowEntities[entity];
      const key = getEntityKey(entityObj);
      if (!entityMaps[entity]) entityMaps[entity] = new Map();
      if (!entityMaps[entity].has(key)) {
        entityMaps[entity].set(key, entityObj);
      }
      if (!topObj) {
        topObj = entityObj; // first entity becomes top-level
      } else {
        // If entity already exists on topObj
        if (!topObj[entity]) {
          topObj[entity] = [];
        }
        // Only push if not already in array
        if (!topObj[entity].some((o: any) => getEntityKey(o) === key)) {
          topObj[entity].push(entityObj);
        }
      }
    }
    if (topObj && !result.includes(topObj)) {
      result.push(topObj);
    }
  }
  return result as T[];
}
function match(
  router: any,
  path: string,
  request: Request,
  env: Env
): Response {
  const segments = path.split("/").filter(Boolean);
  const params: string[] = [];
  let node: any = router;
  const notFound = () =>
    new Response(JSON.stringify({ error: "Route not found", path }), {
      status: 404,
      headers: { "Content-Type": "application/json" },
    });
  
  for (let i = 0; i < segments.length; i++) {
    const segment = segments[i];
    
    if (node[segment]) {
      node = node[segment];
      continue;
    }
    
    // Check for parameterized route (e.g., <id>)
    const paramKey = Object.keys(node).find(
      (k) => k.startsWith("<") && k.endsWith(">")
    );
    
    if (!paramKey) return notFound();
    
    params.push(segment);
    node = node[paramKey];
  }
  
  return typeof node === "function"
    ? node(...params, request, env)
    : notFound();
}

            const $: any = {};

            
                $.ExampleModel = {
                    validate(obj: any): boolean {
                        if (obj.id == null || typeof obj.id !== "number") {return false;}
if (obj.rating == null || typeof obj.rating !== "number") {return false;}
if (obj.description !== undefined && typeof obj.description !== "string") {return false;}
if (obj.data_blob == null || !(obj.data_blob instanceof ArrayBuffer || obj.data_blob instanceof Uint8Array)) {return false;}
                        return true;
                    }
                };
            
        

const router = { api: {
ExampleModel: {create_example: async (request: Request, env: Env) => {
                
if (request.method !== "POST") {
    return new Response("Method Not Allowed", { status: 405 });
}

                
                let body;
                try {
                    body = await request.json();
                } catch {
                    
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
        
                }

                let {payload} = body;
                
if ((payload == null || !$.ExampleModel.validate(payload))) { 
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
         }
payload = Object.assign(new ExampleModel(), payload)
                
                
return JSON.stringify(ExampleModel.create_example(env.DB, payload));

            },
read_example: async (request: Request, env: Env) => {
                
if (request.method !== "GET") {
    return new Response("Method Not Allowed", { status: 405 });
}

                
                let body;
                try {
                    body = await request.json();
                } catch {
                    
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
        
                }

                let {id} = body;
                
if ((id !== undefined && typeof id !== "number")) { 
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
         }
                
                
return JSON.stringify(ExampleModel.read_example(env.DB, id));

            },
list_examples: async (request: Request, env: Env) => {
                
if (request.method !== "GET") {
    return new Response("Method Not Allowed", { status: 405 });
}

                
                
                
return JSON.stringify(ExampleModel.list_examples(env.DB));

            },
update_example: async (request: Request, env: Env) => {
                
if (request.method !== "PUT") {
    return new Response("Method Not Allowed", { status: 405 });
}

                
                let body;
                try {
                    body = await request.json();
                } catch {
                    
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
        
                }

                let {id,rating} = body;
                
if ((id == null || typeof id !== "number")) { 
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
         }
if ((rating !== undefined && typeof rating !== "number")) { 
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
         }
                
                
return JSON.stringify(ExampleModel.update_example(env.DB, id, rating));

            },
patch_example: async (request: Request, env: Env) => {
                
if (request.method !== "PATCH") {
    return new Response("Method Not Allowed", { status: 405 });
}

                
                let body;
                try {
                    body = await request.json();
                } catch {
                    
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
        
                }

                let {description} = body;
                
if ((description !== undefined && typeof description !== "string")) { 
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
         }
                
                
return JSON.stringify(ExampleModel.patch_example(env.DB, description));

            },
delete_example: async (request: Request, env: Env) => {
                
if (request.method !== "DELETE") {
    return new Response("Method Not Allowed", { status: 405 });
}

                
                let body;
                try {
                    body = await request.json();
                } catch {
                    
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
        
                }

                let {id} = body;
                
if ((id == null || typeof id !== "number")) { 
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
         }
                
                
return JSON.stringify(ExampleModel.delete_example(env.DB, id));

            },
process_model: async (request: Request, env: Env) => {
                
if (request.method !== "POST") {
    return new Response("Method Not Allowed", { status: 405 });
}

                
                let body;
                try {
                    body = await request.json();
                } catch {
                    
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
        
                }

                let {model} = body;
                
if ((model == null || !$.ExampleModel.validate(model))) { 
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
         }
model = Object.assign(new ExampleModel(), model)
                
                
return JSON.stringify(ExampleModel.process_model(env.DB, model));

            },
process_models: async (request: Request, env: Env) => {
                
if (request.method !== "POST") {
    return new Response("Method Not Allowed", { status: 405 });
}

                
                let body;
                try {
                    body = await request.json();
                } catch {
                    
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
        
                }

                let {models} = body;
                
if ((models == null || !Array.isArray(models) || models.some(item => (item == null || !$.ExampleModel.validate(item))))) { 
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
         }
models = models.map(item => Object.assign(new ExampleModel(), item))
                
                
return JSON.stringify(ExampleModel.process_models(env.DB, models));

            },
"<id>": { get_description: async (id: number, request: Request, env: Env) => {
                
if (request.method !== "GET") {
    return new Response("Method Not Allowed", { status: 405 });
}

                
                
const d1 = env.DB;
const query = `SELECT * FROM ExampleModel WHERE id = ?`;
const record = await d1.prepare(query).bind(id).first();
if (!record) {
    return new Response(
        JSON.stringify({ error: "Record not found" }),
        { status: 404, headers: { "Content-Type": "application/json" } }
    );
}
const instance = Object.assign(new ExampleModel(), record);

                
return JSON.stringify(instance.get_description(env.DB));

            },
increment_rating: async (id: number, request: Request, env: Env) => {
                
if (request.method !== "PATCH") {
    return new Response("Method Not Allowed", { status: 405 });
}

                
                let body;
                try {
                    body = await request.json();
                } catch {
                    
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
        
                }

                let {amount} = body;
                
if ((amount == null || typeof amount !== "number")) { 
            return new Response(JSON.stringify({ error: "Invalid request body" }), {
                status: 400,
                headers: { "Content-Type": "application/json" },
            });
         }
                
const d1 = env.DB;
const query = `SELECT * FROM ExampleModel WHERE id = ?`;
const record = await d1.prepare(query).bind(id).first();
if (!record) {
    return new Response(
        JSON.stringify({ error: "Record not found" }),
        { status: 404, headers: { "Content-Type": "application/json" } }
    );
}
const instance = Object.assign(new ExampleModel(), record);

                
return JSON.stringify(instance.increment_rating(env.DB, amount));

            } }}
} }


export default {
    async fetch(request: Request, env: Env, ctx: any): Promise<Response> {
        try {
            const url = new URL(request.url);
            return await match(router, url.pathname, request, env);
        } catch (error: any) {
            console.error("Internal server error:", error);
            return new Response(JSON.stringify({ error: error?.message }), {
                status: 500,
                headers: { "Content-Type": "application/json" },
            });
        }
    }
};
